### DISCUSSION TITLE
Architecture for multiple servers and 1 prisma instance

### DISCUSSION THREAD

**Discussion by alan345** [prisma/prisma#3667] [Q&A] [answered]
Lets assume we have this architecture
```
Project
----1 Database (RDS)
----EC2 (AWS)
--------PM2 (multiple servers run by https://github.com/Unitech/pm2 in production)
------------Server1 (main App) 
----------------index1.ts
----------------node_module
----------------prisma
--------------------.env
--------------------schema.prisma (Connect to Database RDS)
------------Server2 (CRON App)
----------------index2.ts
----------------node_module
```

Server1 has acces to the database via Prisma.
How Server2 should access to Prisma? (and the database)

Should I copy and paste  the `schema.prisma` into Server2 and regenerate?
maybe we can isolate `prisma` and save it into the same level as `Server1` and `Server2`?

What is your recomendation?

**Comment by ryands17** [comment, marked as answer]
Hey @alan345 👋 
There are 2 ways you can go about this.

1. One thing can be done is creating a monorepo and create a separate package just that has your schema and Prisma Client.

2. The other would be creating a separate private package with your `prisma.schema`  and migrations and then install that package in your current projects.

In both of these options, you can use the same exported instance of Prisma. Let me know if you have any queries.

**Reply by alan345** [reply]
thanks for your answer. 
by "create a separate package" you mean doing something like:

```
----Prisma (main App) 
--------node_module
--------prisma
------------.env
------------schema.prisma (Connect to Database RDS)
----Server1 (main App) 
--------index1.ts
--------node_module
----Server2 (CRON App)
--------index2.ts
--------node_module
```

if yes, in `index1.ts`, how `import { PrismaClient } from '@prisma/client'` will refer to the `node_module` in `Prisma` folder?

**Reply by ryands17** [reply]
In that separate _Prisma_ package, create an `index.ts` and initialize PrismaClient. As you are creating a separate package, install that in your other projects and use that.

Suppose your `index.ts` in your Prisma app is something like this:

```ts
import { PrismaClient } from '@prisma/client';

export const client = new PrismaClient();
```

Importing this package in your other project:

```ts
import { client } from '@package/prisma';

// use client here
```

**Reply by alan345** [reply]
can you share a link (tuto/doc) about creating a "separate package" to make sure we are in the same page? Everything I am finding is about "creating and publishing a NPM package"

**Reply by ryands17** [reply]
Apologies if I came out wrong. I was suggesting to create an NPM package in my second point.

**Reply by alan345** [reply]
I see. Can you give me more details on your first point? You still talk about "separate package". It is still NPM?

**Reply by ryands17** [reply]
The first pattern is using a monorepo. This setup helps you add a specific folder as a package and you can just add it as a dependency and import that package anywhere. You can check out [this](https://www.smashingmagazine.com/2019/07/yarn-workspaces-organize-project-codebase-pro/) link for how to get started.

Also this is not NPM, you can use your folder locally as a package.

**Reply by alan345** [reply]
thanks!
I was able to to something with 2 folders `prisma` and `server` using `yarn Workspaces `
```
// prisma/index.ts

import { PrismaClient } from '@prisma/client'
export default PrismaClient
```

and 
```
// server/index.ts

import PrismaClient from 'prisma'
const prisma = new PrismaClient()
```

it works well!


But I also need to have access to the type `Invoice` generated by Prisma client. `import { Invoice } from '@prisma/client'`
so I did something like:

```
// prisma/index.ts

import prisma from '@prisma/client'
export default prisma
```

and 
```
// server/index.ts

import allPrisma from 'prisma'

const { PrismaClient } = allPrisma
const prisma = new PrismaClient()
```

Unfortunately all types get lost

![image](https://user-images.githubusercontent.com/15246526/93370262-63bbd300-f805-11ea-8441-47e157c6f260.png)

Any recommendation?

**Reply by ryands17** [reply]
I have created a repo for you so that you can look into it. It's a basic monorepo where I have imported types from Prisma in my frontend app. The same can be done for the backend as well. There are two folders, _client_ and _server_. I have installed the server package from the workspace I am importing a file from `prisma/types.ts` from the server. In the client folder, in `src/App.tsx` I have imported the same type and it works fine.

[prisma-workspace.zip](https://github.com/prisma/prisma/files/5237379/prisma-workspace.zip)

Let me know if you still face any issues.

**Reply by alan345** [reply]
Thanks so much for your detailed help! it works great!

Also I realized that with Yarn workspace (https://classic.yarnpkg.com/en/docs/workspaces/), because you dont have `node_modules` per folder, but only one `node_modules` in the root, shared with all folders, using `import { PrismaClient } from '@prisma/client'` everywhere works because it will hit the shared `node_modules`.

**Comment by ludralph** [comment]
Hi there,

To keep our discussions organized and focused on the most relevant topics, we’re reviewing and tidying up our backlog. As part of this process, we’re closing discussions that have already been marked as answered but remain open.

If this discussion still requires further input or clarification, feel free to reopen it or start a new one with updated details. Your contributions are invaluable to the community, and we’re here to help!

For more details about our priorities and vision for the future of Prisma ORM, check out our latest blog post: https://www.prisma.io/blog/prisma-orm-manifesto.

Thank you for your understanding and ongoing support of the Prisma community!
